plugins/timesync.koplugin/main.lua

```diff
@@ -1,12 +1,14 @@
 local Device = require("device")
+local lfs = require("libs/libkoreader-lfs")
 
-local command
---- @todo (hzj-jie): Does pocketbook provide ntpdate?
-if Device:isKobo() then
-    command = "ntpd -q -n -p pool.ntp.org"
-elseif Device:isCervantes() or Device:isKindle() or Device:isPocketBook() then
-    command = "ntpdate pool.ntp.org"
-else
+local ntp_cmd
+-- Check if we have access to ntpd or ntpdate
+if os.execute("command -v ntpd >/dev/null") == 0 then
```

Comment:
Does a positive response here imply you have the permissions to do what follows? Though I don't know if that was the thought behind the original whitelist approach.

Reply:
No, just that the binary exists. The code does check the actual return value later on, but assumes network issues rather than permissions.

Does setting the time actually require a specific cap?

Reply:
Yeah, `CAP_SYS_TIME`.

Okay, so, PB is probably out then ;p.

Reply:
Added a root check ;).

Reply:
Unfortunately it looks like capget requires syscalls, 'cause of course that's not entirely accurate. (But probably better than the other away around.)

Incidentally this is what I got out of massaging an AI for a couple of minutes. It looks quite plausible, and perhaps with https://github.com/justincormack/ljsyscall it would work.

```lua
local ffi = require "ffi"
local syscall = require "syscall"

ffi.cdef[[
    typedef struct __user_cap_header_struct {
        __u32 version;
        int pid;
    } cap_user_header_t;

    typedef struct __user_cap_data_struct {
        __u32 effective;
        __u32 permitted;
        __u32 inheritable;
    } cap_user_data_t;

    int capget(cap_user_header_t, cap_user_data_t);
]]

local header = ffi.new("cap_user_header_t", { version = 0x20080522, pid = 0 })
local data = ffi.new("cap_user_data_t")

local ret = syscall.capget(header, data)
if ret == 0 then
    if bit.band(data.permitted, 0x08) ~= 0 then
        print("CAP_SYS_TIME capability is present")
    else
        print("CAP_SYS_TIME capability is not present")
    end
else
    print("Error calling capget")
end
```

Reply:
I'm not actually sure we can safely rely on caps instead of full root on most of our target platforms, though (because old kernels & userland).

(But, mostly, we're either already root, or we're not and they don't really have a way to manipulate caps anyway ;p).

Good to know, though ;).